### DISCUSSION TITLE
how to perform e-2-e tests which requires login

### DISCUSSION THREAD

**Discussion by bertuz** [leancodepl/patrol#1604] [Q&A]
I'm developing a flutter application which interacts with a supabase backend for both API and Auth. The only authentication is via email, specifically by means of magic link.

The problem: I want to write some end-to-end test, and to do that I'm using patrol. The test built are saved on a public repo along with the codebase, in order to perform some github actions which run the tests on firebase.
Of course I cannot log into the app via the magic link: that would mean opening the web browser, having the email inbox set on the virtual device etc.

The possible solution: at patrol build time, set environment variables read by `String.fromEnvironment` in order to use user+password credentials login when testing.

pros:
- it works

cons:
- This means extra code - although very little - that will only exists for testing purposes: checking if it's running with tests env vars and using the email+password way to log in
 - The compiled patrol outputs will have test credentials which are really easy to extract. Although it's using a test environment, this should be avoided, because I compile the apk and iOS packages locally and they are sent to a public repo in order to save github actions' resources. 

I don't know any other way to accomplish this, if not mocking all the supabase's calls. What could I do?

**Comment by bartekpacia** [comment]
You can use a mock email service like 1secmail (or similar) and use their API to extract emails contents and go to the magic link.

@fylyppo I believe you encountered this use case, could you share your learnings?

**Comment by fylyppo** [comment]
Hi, I don't get your authentication flow.
If you need to get the magic link from email you can use https://www.1secmail.com/api/. It was reliable solution for me - I was doing the same as you, I needed to open this link to authorize myself.
If you need to use credentials (user+password) then use `String.fromEnvironment` and [Github Actions Secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets).

**Reply by bertuz** [reply]
Nice and clean, thanks for the idea!
